minerl/herobraine/hero/handlers/agent/observations/location_stats.py

```diff
@@ -0,0 +1,158 @@
+                _XPositionObservation(),
+                _YPositionObservation(),
+                _ZPositionObservation(),
+                _SeaLevelObservation()
+            ]
+        )
+
+
+class _FullStatsObservation(KeymapTranslationHandler):
+    def to_hero(self, x) -> int:
+        for key in self.hero_keys:
+            x = x[key]
+        return x
+
+    def __init__(self, key_list : List[str], space=None, default_if_missing=None):
+        if space is None:
+            if 'achievement' == key_list[0]:
+                space = spaces.Box(low=0, high=1, shape=(), dtype=np.int)
+            else:
+                space = spaces.Box(low=0, high=np.inf, shape=(), dtype=np.int)
+        if default_if_missing is None:
+            default_if_missing = np.zeros((),dtype=np.float)
+
+
+        super().__init__(hero_keys=key_list, univ_keys=key_list, space=space,
+                         default_if_missing=default_if_missing)
+
+    def xml_template(self) -> str:
+        return str("""<ObservationFromFullStats/>""")
+
+
+class _SunBrightnessObservation(_FullStatsObservation):
+    def __init__(self):
+        super().__init__(key_list=['sun_brightness'], space=spaces.Box(low=0.0, high=1.0, shape=(), dtype=np.float), default_if_missing=0.94)
+
+class _SkyLightLevelObservation(_FullStatsObservation):
+    def __init__(self):
+        super().__init__(key_list=['sky_light_level'], space=spaces.Box(low=0.0, high=1.0, shape=(), dtype=np.float), default_if_missing=0.71)
+
+class _XPositionObservation(_FullStatsObservation):
+    def __init__(self):
+        super().__init__(key_list=['xpos'], space=spaces.Box(low=-640000.0, high=640000.0, shape=(), dtype=np.float), default_if_missing=0.0)
+class _YPositionObservation(_FullStatsObservation):
+    def __init__(self):
+        super().__init__(key_list=['ypos'], space=spaces.Box(low=-640000.0, high=640000.0, shape=(), dtype=np.float), default_if_missing=0.0)
+class _ZPositionObservation(_FullStatsObservation):
+    def __init__(self):
+        super().__init__(key_list=['zpos'], space=spaces.Box(low=-640000.0, high=640000.0, shape=(), dtype=np.float), default_if_missing=0.0)
+
+class _PitchObservation(_FullStatsObservation):
+    def __init__(self):
+        super().__init__(key_list=['pitch'], space=spaces.Box(low=-180.0, high=180.0, shape=(), dtype=np.float),
+                         default_if_missing=0.0)
+
+class _YawObservation(_FullStatsObservation):
+    def __init__(self):
+        super().__init__(key_list=['yaw'], space=spaces.Box(low=-180.0, high=180.0, shape=(), dtype=np.float),
+                         default_if_missing=0.0)
+
+class _BiomeIDObservation(_FullStatsObservation):
+    def to_hero(self, x):
+        for key in self.hero_keys:
+            x = x[key]
+        return np.eye(40)[x]
+
+    def __init__(self):
+        super().__init__(key_list=['biome_id'],
+                         space=spaces.Discrete(40),
+                         default_if_missing=np.eye(40)[0])
```

Comment:
Is it because there are 40 biomes? Is there a good way to handle a change in the number of biomes?

Reply:
Also, probably not a big deal but unless I'm mistaken this creates an entire 40x40 matrix to only extract the x'th row. Is there a standard way to do this (a onehot function we already have on hand perhaps) that doesn't allocate and initialize 40 times more data than it needs to?

Reply:
You're correct, I stole this from another project. It does use more space but it is fewer lines than making a vector and assigning the nth element to 1. Since there could be more biomes later I think I will just refactor all of these to be int. In general we preferred showing the independence of consecutive values by encoding them as a one-hot from MineRL because not everyone wraps things.

Reply:
Nevermind - the discrete space dictates the type here. There are currently biomes with IDs as large as ~400 ... they don't appear in the base game but could be enabled later